### ISSUE TITLE
Provide an H5P caretaker service for authenticated users on a hosted PB network

### ISSUE THREAD

**Issue by SteelWagstaff** [pressbooks/ideas#542]
Earlier in 2025, NDLA/Oliver Tacke released a ‘H5P Caretaker’ tool that allows people to create a page where users can upload h5p files and have them tested for a number of different accessibility/usability issues. Users can view a report and fix any issues flagged. Reference implementation here: https://explore.ndla.no/h5p-caretaker/. This is available as a WordPress plugin: https://wordpress.org/plugins/ndla-h5p-caretaker/. 

This might be nice to provide to our clients – rather than hosting a public version, one idea might be to activate on the root site of each enterprise network and make available to anyone with a user account so that they could check their h5p files at https://MYURL.network/h5p-caretaker.

Oliver Tacke writes:
> The Caretaker "architecture" was designed with flexibility in mind. There's the (PHP based) server-side library that does all the heavy lifting, there's a client-side app that you can use as is or use the information that you receive from the server-side
as you please. And there are different integrations bringing both together, including [a WordPress plugin](https://wordpress.org/plugins/ndla-h5p-caretaker/) that allows you to get a report for the content that you are viewing easily. You just need to replace what doesn't suit your needs while keeping the rest.

## Notes from initial testing

To make this work locally, for logged in users, user permissions must be configured according to these instructions:
https://github.com/NDLANO/wp-ndla-h5p-caretaker?tab=readme-ov-file#configure. We'll need to set the use-h5p-caretaker permissions as desired. I have opened a PR in pressbooks-plugins-config that allows any logged in user to user the caretaker page if the plugin is active on the network root site: https://github.com/pressbooks/pressbooks-plugins-config/pull/63.

In June 2025 we installed and enabled the WordPress plugin on dev.pressbooks.pub for initial testing (see https://github.com/pressbooks/private/issues/1763). Because of how we've modified the network menus, the settings page for the plugin is obscured and can only be accessed by direct link (i.e. `https://dev.pressbooks.pub/wp-admin/options-general.php?page=ndlah5pcaretaker-admin`). I left the default URL and added sample header and footer text. I was unable to view the page, however -- got a 500 'This page isn't working error': `https://dev.pressbooks.pub/h5p-caretaker/`. Tried setting different URL and making the page public (i.e. not requiring users to be logged in and given permissions), each with similar results. Christopher confirmed that he also saw a 500 error at the URL and that it shows up in access.log as a 500 error, but nothing shows up in the error.log.

**Comment by SteelWagstaff** [comment]
@amandacoolidge Keep. This is an important complementary component for accessibility reports in Content Toolkit. If enabled, would allow users to perform accessibility checks (and more) for their H5P activities. I researched previously but got blocked by operations issues that were never investigated or resolved.

**Comment by amandacoolidge** [comment]
@SteelWagstaff  @MichSpieler should this be moved to accessibility pipeline?

**Comment by MichSpieler** [comment]
I am willing to accept it there. If it is high in priority, the CFT pipeline might be more promising atm.

**Comment by SteelWagstaff** [comment]
@arzola it would be nice to research this a bit further to find out why it doesn't work in our infrastructure

**Comment by SteelWagstaff** [comment]
@amandacoolidge This is PB enterprise backlog work, related to the content toolkit.

**Comment by cmurtagh** [comment]
Here's the error we get:

```
PHP Fatal error:  Uncaught TypeError: array_map(): Argument #1 ($callback) must be a valid callback or null, class "\Locale" not found in /web/app/plugins/ndla-h5p-caretaker/includes/render-index.php:382
Stack trace:
#0 /web/app/plugins/ndla-h5p-caretaker/includes/render-index.php(382): array_map('\\Locale::getDis...', Array, Array)
#1 /web/app/plugins/ndla-h5p-caretaker/includes/render-index.php(315): NDLAH5PCARETAKER\get_select_language_locales('en_US')
#2 /web/app/plugins/ndla-h5p-caretaker/includes/render-index.php(83): NDLAH5PCARETAKER\render_html(Array)
#3 /web/app/plugins/ndla-h5p-caretaker/includes/class-main.php(168): NDLAH5PCARETAKER\render_page_index()
#4 /web/wp/wp-includes/class-wp-hook.php(341): NDLAH5PCARETAKER\Main->handle_requests('')
#5 /web/wp/wp-includes/class-wp-hook.php(365): WP_Hook->apply_filters(NULL, Array)
#6 /web/wp/wp-includes/plugin.php(522): WP_Hook->do_action(Array)
#7 /web/wp/wp-includes/template-loader.php(23): do_action('template_redire...')
#8 /web/wp/wp-blog-header.php(19): require_once('/srv/www/bedroc...')
#9 /web/index.php(6): require('/srv/www/bedroc...')
#10 {main}
  thrown in /web/app/plugins/ndla-h5p-caretaker/includes/render-index.php on line 382
```

**Comment by SteelWagstaff** [comment]
I think this error message is happening because of a PHP extension: our production server is missing the PHP `intl` extension that our local environments have enabled. Can you try installing/enabling that extension on the dev server @cmurtagh?

**Comment by cmurtagh** [comment]
> I think this error message is happening because of a PHP extension: our production server is missing the PHP `intl` extension that our local environments have enabled. Can you try installing/enabling that extension on the dev server [@cmurtagh](https://github.com/cmurtagh)?

Wow, that's embarrassing, I should have caught that. I'll also check our IaC to ensure that this package is included. Thank you for pointing it out.

**Comment by SteelWagstaff** [comment]
@cmurtagh I'm now able to access the caretaker page on the dev.pressbooks.pub network. However, when I attempt to upload a file to the checker, I see a `Something went wrong, but I dunno what, sorry! Please check the server log.` 

The console displays the following error:

<img width="818" height="145" alt="Image" src="https://github.com/user-attachments/assets/26672e23-f88f-4fad-8dbb-1c62fa1187c3" />

I cannot reproduce the error locally (the tool works as expected).

When I look at the network tab, I can see that browser shows a real POST to https://dev.pressbooks.pub/h5p-caretaker-upload with a 403 Forbidden response. My interpretation is that the failure is happening after the browser sends the request and before the server allows the action. I don't have the tools or server access to investigate or diagnose this further.

**Comment by SteelWagstaff** [comment]
Response headers and body for failed POST action can be provided if needed, or you can probably reproduce in your own browser.

**Comment by cmurtagh** [comment]
@ho-man-chan is it possible that CloudFlare is blocking this?

**Comment by ho-man-chan** [comment]
I have adjusted the security rule. @SteelWagstaff @cmurtagh can try again.

**Comment by SteelWagstaff** [comment]
Works for me. Thanks!

**Comment by cmurtagh** [comment]
Yup, works for me too.

**Comment by MichSpieler** [comment]
From A11y WG meeting: Ideally we would ask Oliver to support us here.
